### PR TITLE
 [WIP] (models) Use different sessions collection per experiment

### PR DESCRIPTION
**Pull request by abought**
Refs https://openscience.atlassian.net/browse/LEI-78

## Purpose
See detailed notes on companion PR, https://github.com/CenterForOpenScience/experimenter/pull/31

### CODE REVIEW

File: exp-models/addon/mixins/jam-document-adapter.js [hunk 1]
```diff
@@ -17,6 +17,7 @@ export default Ember.Mixin.create({
         collectionId: (type /*, id, snapshot, query*/) => Ember.Inflector.inflector.pluralize(type)
     },
     ajax: function(url, type, options) {
```
Comment (samchrisinger) [inline exp-models/addon/mixins/jam-document-adapter.js]: Just fyi es6 supports: `ajax: function(url, type, options={}) {`

File: exp-models/addon/models/experiment.js [hunk 2]
```diff
@@ -21,14 +26,31 @@ export default DS.Model.extend(JamModel, {
...
+
+    _registerSessionModels() {
+        // Dynamically register the required models for a session table associated with this experiment
+        var cId = this.get('sessionCollectionId');
+        window.App.register(`model:${cId}`, SessionModel.extend()); // register a dummy model. This seems to work even if model already registered
+        window.App.register(`adapter:${cId}`, SessionAdapter.extend({sessionCollectionId: cId})); // Override part of adapter URL
+        window.App.register(`serializer:${cId}`, SessionSerializer.extend({modelName: cId})); // Tell serializer what model to use)
+    },
+
+    init() {
```
Comment (samchrisinger) [inline exp-models/addon/models/experiment.js]: Do you need a super call here?

File: exp-models/addon/mixins/jam-document-adapter.js (hunk 1 above)
Comment (abought) [inline exp-models/addon/mixins/jam-document-adapter.js]: O' frabjous day! Callooh! Callay!
(changed this and verified working)

File: exp-models/addon/models/experiment.js (hunk 2 above)
Comment (abought) [inline exp-models/addon/models/experiment.js]: There wasn't a parent method to call- but there might be. Good catch.

File: exp-models/addon/mixins/jam-document-serializer.js [hunk 3]
```diff
@@ -10,7 +10,7 @@ export default Ember.Mixin.create({
         modifiedBy: {serialize: false},
     },
 
-    payloadKeyFromModelName: function(modelName) {
+    payloadKeyFromModelName: function(/*modelName */) {
         // JamDB expects all collections to specify JSONAPI type 'documents'
         return 'documents';
```
Comment (chrisseto) [inline exp-models/addon/mixins/jam-document-serializer.js]: What about collections?

Comment (samchrisinger) [inline exp-models/addon/mixins/jam-document-serializer.js]: see `jam-document-serializer`

File: exp-models/addon/models/experiment.js [hunk 4]
```diff
@@ -21,14 +26,32 @@ export default DS.Model.extend(JamModel, {
...
+    sessionCollectionId: Ember.computed('shortId', function() {
+        // Return a string corresponding to the session collection shortID, to be used by model/adapter/serializer
+        // Eg an experiment called 'test0' would have a collection 'session-test0'
+        return `session${this.get('shortId')}s`; // FIXME: the spurious s is a requirement imposed by genschemas...
+    }),
+
+    _registerSessionModels() {
+        // Dynamically register the required models for a session table associated with this experiment
+        var cId = this.get('sessionCollectionId');
+        window.App.register(`model:${cId}`, SessionModel.extend()); // register a dummy model. This seems to work even if model already registered
```
Comment (chrisseto) [inline exp-models/addon/models/experiment.js]: Is there a better workaround to this?  Using `window` in a built app makes me uneasy

Comment (abought) [inline exp-models/addon/models/experiment.js]: It makes me a little worried too. 

It seems that Ember [removed](http://emberigniter.com/accessing-global-object-in-ember-cli-app/) the global app reference it used to have, and this approach seems to depend on calling `application.register`... I honestly don't know if there's another way to get at the app instance.

This is about all I could find on the lifetime of the window object, but that's a thin shield. I agree I'd like to see a better way.

**Pull request merged**